src/gar.lua

```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
+		return fundingPlan
+	end
+
+	planMinimumStakesDrawdown(fundingPlan, gatewaysInfo)
+
+	return fundingPlan
+end
+
+function planBalanceDrawdown(fundingPlan, sourcesPreference)
+	local availableBalance = balances.getBalance(fundingPlan.address)
+	if sourcesPreference == "balance" or sourcesPreference == "any" then
+		fundingPlan.balance = math.min(availableBalance, fundingPlan.shortfall)
+		fundingPlan.shortfall = fundingPlan.shortfall - fundingPlan.balance
+	end
+end
+
+function planExcessStakesDrawdown(fundingPlan)
+	-- find all the address's delegations across the gateways
+	local gatewaysInfo = utils.sortTableByFields(
+		utils.reduce(
+			-- only consider gateways that have the address as a delegate
+			utils.filterDictionary(gar.getGatewaysUnsafe(), function(_, gateway)
+				return gateway.delegates[fundingPlan.address] ~= nil
+			end),
+			-- extract only the essential gateway fields, copying tables so we don't mutate references
+			function(acc, gatewayAddress, gateway)
+				local totalEpochsGatewayPassed = gateway.stats.passedEpochCount or 0
+				local totalEpochsParticipatedIn = gateway.stats.totalEpochCount or 0
+				local gatewayRewardRatioWeight = (1 + totalEpochsGatewayPassed) / (1 + totalEpochsParticipatedIn)
+				local delegate = utils.deepCopy(gateway.delegates[fundingPlan.address])
+				delegate.excessStake = math.max(0, delegate.delegatedStake - gateway.settings.minDelegatedStake)
+				delegate.gatewayAddress = gatewayAddress
+				table.insert(acc, {
+					totalDelegatedStake = gateway.totalDelegatedStake, -- for comparing gw total stake
+					gatewayRewardRatioWeight = gatewayRewardRatioWeight, -- for comparing gw performance
+					delegate = delegate,
+					startTimestamp = gateway.startTimestamp, -- for comparing gw tenure
+				})
+				return acc
+			end,
+			{}
+		),
+		{
+			{
+				order = "desc",
+				field = "delegate.excessStake",
+			},
+			{
+				order = "asc",
+				field = "gatewayRewardRatioWeight",
+			},
+			{
+				order = "desc",
+				field = "totalDelegatedStake",
+			},
+			{
+				order = "desc",
+				field = "startTimestamp",
+			},
+		}
+	)
+
+	-- simulate drawing down excess stakes until the remaining balance is satisfied OR excess stakes are exhausted
+	local gatewayIndex, nextGateway = next(gatewaysInfo)
+	while fundingPlan.shortfall > 0 and nextGateway do
+		local excessStake = nextGateway.delegate.excessStake
+		local stakeToDraw = math.min(excessStake, fundingPlan.shortfall)
+		fundingPlan["stakes"][nextGateway.delegate.gatewayAddress] = {
+			delegatedStake = stakeToDraw,
+			vaults = {}, -- set up vault spend tracking now while we're passing through
+		}
+		fundingPlan.shortfall = fundingPlan.shortfall - stakeToDraw
+		nextGateway.delegate.delegatedStake = nextGateway.delegate.delegatedStake - stakeToDraw
+		-- maintain consistency
+		nextGateway.delegate.excessStake = excessStake - stakeToDraw
+		nextGateway.totalDelegatedStake = nextGateway.totalDelegatedStake - stakeToDraw
```

Comment:
can you explain why this? are you updating the reference to the current gateway to avoid double counting the stake available from it? more detailed comment might be helpful

Reply:
ahhh i see - makes sense

Reply:
Expanded comment:
98f6b38